server/remote_cache/cachetools/cachetools.go

```diff
@@ -269,6 +276,232 @@ func UploadProtoToCAS(ctx context.Context, cache interfaces.Cache, instanceName
 	return uploadProtoToCache(ctx, cas, instanceName, in)
 }
 
+// BatchCASUploader uploads many files to CAS concurrently, batching small
+// uploads together and falling back to bytestream uploads for large files.
+type BatchCASUploader struct {
+	ctx              context.Context
+	eg               *errgroup.Group
+	cache            interfaces.Cache
+	bsClient         bspb.ByteStreamClient
+	casClient        repb.ContentAddressableStorageClient
+	req              *repb.BatchUpdateBlobsRequest
+	currentBatchSize int64
+	instanceName     string
+	uploads          map[digest.Key]struct{}
+}
+
+func NewBatchCASUploader(ctx context.Context, env environment.Env, instanceName string) (*BatchCASUploader, error) {
+	bsClient := env.GetByteStreamClient()
+	if bsClient == nil {
+		return nil, status.InvalidArgumentError("Missing bytestream client")
+	}
+	casClient := env.GetContentAddressableStorageClient()
+	if casClient == nil {
+		return nil, status.InvalidArgumentError("Missing CAS client")
+	}
+	cache := env.GetCache()
+	if cache == nil {
+		return nil, status.InvalidArgumentError("Missing cache")
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
```

Comment:
errorgroup cancels "ctx" as soon as any error is returned from a closure or Wait returns. Just want to make sure that fits with your mental model here and isn't going to prematurely cancel something.